Core/NoOpRunner.Core/Shapes/BaseShape.cs

```diff
@@ -74,5 +74,21 @@ public virtual void OnClick()
 
         public abstract bool CanOverlap(BaseShape other);
         public abstract void OnCollision(BaseShape other);
+
+        public BaseShape Clone()
+        {
+            return (BaseShape)this.MemberwiseClone();
+        }
+        public BaseShape DeepCopy()
+        {
+            BaseShape clone = (BaseShape)this.MemberwiseClone();
+            clone.ShapeBlocks = new List<ShapeBlock>(ShapeBlocks);
+            return clone;
```

Comment:
This is not a deep copy. The copy will share the same references to ShapeBlocks. You can check this by creating a copy of a BaseShape, and then modifying an offset of a shapeBlock from the original shape. The "deep" copy will also reflect the changes: 
```
BaseShape deep = shape.DeepCopy();
var sp = shape.GetShapes().First();
var sp1 = deep.GetShapes().First();
sp.OffsetX = 55;
var trueDeep = sp1.OffsetX != sp.OffsetX;```

Reply:
Fixed